### PR TITLE
Toggle profile globally

### LINKED ISSUE #30866: User Profile - 3 - Global toggle
- Disable/enable profile feature globally in `config.php` and let admins set the default state which can be overridden by users https://github.com/nextcloud/server/pull/29372

### PR DESCRIPTION
**Pull request by Pytal**
Adds two new settings:
1. A `config.php` option `'profile.enabled'` which defaults to true to enable/disable profile globally
2. A Profile section under Administration “Basic settings” to enable/disable profile by default for new users

Close https://github.com/nextcloud/server/issues/30866

Contributes to https://github.com/nextcloud/server/issues/28139

### CODE REVIEW

**Comment by Pytal** [comment]
@jolly-jump in the current state of this PR, setting the option to disabled would disable profile completely and users would not be able to enable their own profile individually.

You've made a good suggestion so we will also add an admin option to set profile to enabled/disabled by default for all users and allow individual users to opt-in or opt-out.

**Comment by shieldwed** [comment]
For those searching for a quick fix to disable all users profiles for now:
```sql
update oc_accounts set data = JSON_SET(data, '$.profile_enabled', JSON_QUERY('{"value": "0", "scope": "v2-local", "verified": "0"}', '$'));
```

**Comment by kroerig** [comment]
Any idea when the feature will be ready to use? We use NC at our school und students must not be able to enable a public profile.

I applied the workaround from @shieldwed , but this only applies to existing users.

**Comment by skjnldsv** [comment]
tests failure

**Review by szaimen — APPROVED** [review]
The failing test seems unrelated, afaics.
I tested this and it seems to work fine 👍
I haven't had a look at the code though.

File: lib/private/Profile/TProfileHelper.php [hunk 1]
```diff
@@ -26,19 +26,12 @@
...
-	/**
-	 * Returns whether the profile is enabled for the account
-	 *
-	 * @since 23.0.0
-	 */
-	protected function isProfileEnabled(IAccount $account): ?bool {
+	protected function isProfileEnabledByDefault(IConfig $config): ?bool {
 		return filter_var(
-			$account->getProperty(IAccountManager::PROPERTY_PROFILE_ENABLED)->getValue(),
+			$config->getAppValue('settings', 'profile_enabled_by_default', '1'),
```
Comment (szaimen) [inline lib/private/Profile/TProfileHelper.php:34]: maybe we should document this somewhere?

File: lib/private/Profile/TProfileHelper.php [hunk 2]
```diff
@@ -26,19 +26,12 @@
...
 
 trait TProfileHelper {
-
-	/**
-	 * Returns whether the profile is enabled for the account
-	 *
-	 * @since 23.0.0
-	 */
-	protected function isProfileEnabled(IAccount $account): ?bool {
+	protected function isProfileEnabledByDefault(IConfig $config): ?bool {
```
Comment (artonge) [inline lib/private/Profile/TProfileHelper.php:32]: Will always be a bool, no?

```suggestion
	protected function isProfileEnabledByDefault(IConfig $config): bool {
```

File: lib/private/Profile/ProfileManager.php [hunk 3]
```diff
@@ -123,6 +129,24 @@ public function __construct(
 		$this->coordinator = $coordinator;
 	}
 
+	/**
+	 * If no user is passed as an argument return whether profile is enabled globally in `config.php`
+	 */
+	public function isProfileEnabled(?IUser $user = null): ?bool {
```
Comment (artonge) [inline lib/private/Profile/ProfileManager.php:135]: Will always be a bool, no?

```suggestion
	public function isProfileEnabled(?IUser $user = null): bool {
```

File: lib/private/Profile/TProfileHelper.php (hunk 1 above)
Comment (Pytal) [reply lib/private/Profile/TProfileHelper.php:34]: > maybe we should document this somewhere?

Created a separate issue https://github.com/nextcloud/documentation/issues/8122

File: lib/private/Profile/TProfileHelper.php (hunk 2 above)
Comment (Pytal) [reply lib/private/Profile/TProfileHelper.php:32]: `FILTER_NULL_ON_FAILURE` is used for `filter_var()` to be more strict though only truthiness is checked when this is called so this is mainly to distinguish between expected and unexpected values for debugging purposes should it be needed in the future

File: lib/private/Profile/TProfileHelper.php (hunk 1 above)
Comment (jancborchardt) [reply lib/private/Profile/TProfileHelper.php:34]: Just for the record from what we talked about in the chat:
- We are keeping Profile enabled by default
- Admins can easily disable it with this setting by default, with users still being able to override it
- New profile fields which are sensitive were a) introduced recently together with the visibility settings and b) not public by default, so if people don't want them to be visible, they didnt set them

File: lib/private/Profile/TProfileHelper.php (hunk 2 above)
Comment (artonge) [reply lib/private/Profile/TProfileHelper.php:32]: Not sure that I understand the rational, but ok, I trust your judgment ;)

**Review by artonge — APPROVED** [review]

**Comment by szaimen** [comment]
@Pytal can you please rebase? Then lets merge this, I'd say

**Comment by szaimen** [comment]
tests are still failing :/

**Comment by Pytal** [comment]
Looks unrelated but will rerun for safety

**Comment by Pytal** [comment]
🟢

**Pull request merged**